wgsl/index.bs

```diff
@@ -6983,11 +6981,283 @@ TODO: *Stub*: Expression evaluation
+    - If there was a path from CF_return to CannotBeUniform, add an edge from CF_after to CannotBeUniform
+    - If there was a path from Value_return to CannotBeUniform, add an edge from Result to CannotBeUniform
+- Similarly for each argument *i*:
+    - If there was a path from MustBeUniform to arg_i, add an edge from MustBeUniform to arg_i
+    - If there was a path from CF_return to arg_i, add an edge from CF_after to arg_i
+    - If there was a path from Value_return to arg_i, add an edge from Result to arg_i
+- Add an edge from Result to CF_after, and an edge from CF_after to the last CF_i
+- Finally return the nodes "CF_after" and "Result"
+
+Note: Notice that this rule only requires adding a number of edges bounded by 1 + the number of parameters of the functions, independently of how complex the implementation of the function might be. This is key to the linear complexity of the overall algorithm.
+
+Functions in the standard library count as having a path from "Value_return" to all their "arg_i" and no other relevant paths to remember. Here is the list of exceptions to this rule:
+- The `||` and `&&` operators. Their rules are given in the next section (they cannot be treated like function calls because of short-circuiting)
+- TODO
+
+### Uniformity rules for expressions ### {#uniformity-expressions}
+
+The rules for analyzing expressions take as argument both the expression itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A node corresponding to control flow at the exit of it
+* A node corresponding to its value
+* A set of new nodes and edges to add to the graph
+
+Like for statements, if the set of behaviors for an expression is a singleton, then the node corresponding to control flow at the exit of it is replaced by *CF* (the control flow at the beginning of it), overriding the rules below.
+
+<table class='data'>
+  <caption>Uniformity rules for expressions (in normal rvalue position)</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node, value node<th>New edges
+  </thead>
+  <tr><td class="nowrap">e1 || e2
+      <td>
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*<br>
+          *V1*, *e2* -> *CF2*, *V2*
+      <td>*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">e1 && e2
+      <td>
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*<br>
+          *V1*, *e2* -> *CF2*, *V2*, *B2*
+      <td>*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">Literal
+      <td>
+      <td>
+      <td>*CF*, *CF*
+      <td>
+   <tr><td class="nowrap">reference to local variable "x"
+      <td>*Result*
+      <td class="nowrap">*X* is the node corresponding to "x"
+      <td>*CF*, *Result*
+      <td class="nowrap">*Result* -> {*CF*, *X*}
+   <tr><td class="nowrap">reference to global variable "x"
+      <td>
+      <td>
+      <td>*CF*, *CannotBeUniform*
```

Comment:
I think we should split the building of the expression for the reference and the dereferencing steps. This would allow computing that a reference is uniform, such that we can have a builtin that helps bypass the limitation that you can never get uniform values from memory (storage or workgroup).

`[[uniform]] fn uniformLoad(p: [[uniform]] ptr<S, T>) -> [[uniform]] T`

This means that uniformLoad must be called in uniform CF with a uniform `p` reference/pointer (I'm not up to date with that part of the spec), and returns to you a uniform value. This would have a special rule in the uniformity analysis.

Reply:
Unfortunately, a pointer being uniform does not by itself guarantee that values loaded from it will be uniform, as different threads may have written different values to the underlying memory before/through a different pointers.

There are two solutions to this:
- loading from variables in a given storage class (e.g. workgroup) is automatically uniform at the same scope (e.g. workgroup-uniform) as long as the control-flow is also (workgroup-)uniform. This does not require any annotation, and would be a straightforward extension of the model.
- For all other cases, the variable itself would have to be annotated; then every store to that variable would have to prove that it is of a uniform value in uniform control-flow, and every load of the variable in uniform control-flow would be able to assume that its result is uniform. I intend to propose this kind of annotation in a follow-up PR after some discussions with Dzmitry.

In both cases, I don't see an advantage to separating the load from the referencing of the variable.

Reply:
> loading from variables in a given storage class (e.g. workgroup) is automatically uniform at the same scope (e.g. workgroup-uniform) as long as the control-flow is also (workgroup-)uniform. This does not require any annotation, and would be a straightforward extension of the model.

This is the most similar to what I was thinking. But here what a developer would want to do is: in uniform control flow, with a uniform pointer to `storage`, load the value so that it is uniform across the whole workgroup. I think it would require the split because the reference is uniform, but the normal load of the variable is not.